test/mapmatch.cc

```diff
@@ -1232,6 +1232,33 @@ TEST(Mapmatch, test_loop_matching) {
   }
 }
 
+TEST(Mapmatch, test_loop_matching_walk_or_snap) {
+  // The walk_or_snap case sometimes incorrectly identifies trivial cases where it can optimize away.
+  // This tests loops which are not trivial.
+  std::vector<std::string> test_cases = {
+      // Simple loop around a block; incorrectly identified as trivial in such a way that
+      // it caused an exception in TripLegBuilder (no maneuvers)
+      R"({"shape_match":"walk_or_snap", "shape":[
+          {"lat": 52.11870746741026, "lon": 5.117783546447755, "type": "break", "node_snap_tolerance": 0},
+          {"lat": 52.11905661947785, "lon": 5.118234157562257, "type": "via", "node_snap_tolerance": 0},
+          {"lat": 52.11934977334695, "lon": 5.117751359939576, "type": "via", "node_snap_tolerance": 0},
+          {"lat": 52.118852398789194, "lon": 5.116935968399049, "type": "via", "node_snap_tolerance": 0},
+          {"lat": 52.11852959643748, "lon": 5.117445588111878, "type": "via", "node_snap_tolerance": 0},
+          {"lat": 52.11870746741026, "lon": 5.117783546447755, "type": "break", "node_snap_tolerance": 0}],
+          "costing":"auto"})",
+  };
+
+  api_tester tester;
+  for (size_t i = 0; i < test_cases.size(); ++i) {
+    // Verify that the map_snap output matches the walk_or_snap output.
+    // This approach is inspired by the above.
```

Comment:
Quick note in case it sparks any ideas... I attempted the same logic as above with verifying that a route over the same points was identical, but the /route response always ended up with a spurious keep right + continue after the 2nd maneuver. Figuring that out is out of scope of this PR, so I went with something a bit more simple. We expect the `map_snap` and `walk_or_snap` results to match, where they did not before.

Reply:
we'd need to look at the trip path. there has to be a difference in how the edges look like two copies of the same edge where a via point lands for example. its unfortunate that that would happen but i totally could see that it could happen.